File: src/main/java/io/github/shaksternano/mediamanipulator/media/ImageFrame.java
package io.github.shaksternano.mediamanipulator.media;

import java.awt.image.BufferedImage;
import java.util.Objects;

public record ImageFrame(
    BufferedImage content,
    double duration,
    long timestamp
) implements VideoFrame<BufferedImage> {

    @Override
    public ImageFrame transform(float speedMultiplier) {
        return transform(content, speedMultiplier);
    }

    @Override
    public ImageFrame transform(BufferedImage newContent, float speedMultiplier) {
        return new ImageFrame(
            newContent,
            duration / speedMultiplier,
            timestamp
        );
    }

    @Override
    public int hashCode() {
        return Objects.hash(
            duration,
            timestamp,
            ImageUtil.hashImage(content)
        );
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        } else if (obj instanceof ImageFrame other) {
            return Double.compare(other.duration, duration) == 0
                && timestamp == other.timestamp
                && ImageUtil.imageEquals(content, other.content());
        } else {
            return false;
        }
    }
}

File: core/src/main/kotlin/io/github/shaksternano/borgar/core/media/ImageFrame.kt
package io.github.shaksternano.borgar.core.media

import java.awt.image.BufferedImage

data class ImageFrame(
    override val content: BufferedImage,
    override val duration: Double,
    override val timestamp: Long
) : VideoFrame<BufferedImage, ImageFrame> {

    override fun transform(newContent: BufferedImage): ImageFrame {
        return transform(newContent, 1F)
    }

    override fun transform(speedMultiplier: Float): ImageFrame {
        return transform(content, speedMultiplier)
    }

    override fun transform(newContent: BufferedImage, speedMultiplier: Float): ImageFrame {
        return ImageFrame(
            newContent,
            duration / speedMultiplier,
            timestamp
        )
    }

    override fun transform(duration: Double, timestamp: Long): ImageFrame {
        return ImageFrame(content, duration, timestamp)
    }
}
